### PR TITLE
split: add app_partition_split request and unit test

### PR DESCRIPTION
**Pull request by hycdong**
app_partition_split is the request which client send to meta server, meta server will validate request and change partition count on remote storage.

refer to #69

### CODE REVIEW

File: include/dsn/dist/replication/replication.types.h [hunk 1]
```diff
@@ -3,84 +3,89 @@
 #include <dsn/service_api_cpp.h>
 #include <dsn/cpp/serialization.h>
 
-namespace dsn {
```
Comment (neverchanje) [inline include/dsn/dist/replication/replication.types.h]: codeformat

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    _meta_svc->get_remote_storage()->set_data(
+                _state->get_app_path(*app),
+                std::move(value),
+                LPC_META_STATE_HIGH,
+                on_write_storage_complete,
+                _meta_svc->tracker());
+}
+
+meta_split_service::meta_split_service(meta_service* meta_srv){
+    this->_meta_svc = meta_srv;
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: 不需要 this->

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    // if init_partition_count is invalid, init_partition_count = original partition_count
+    if (app->init_partition_count < 0){
+        app->init_partition_count = app->partition_count;
+    }
+    auto copy = *app;
+    copy.partition_count *= 2;
+    blob value = dsn::json::json_forwarder<app_info>::encode(copy);
+
+    _meta_svc->get_remote_storage()->set_data(
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: 用 meta storage， 不要用 remote storage

Comment (neverchanje) [reply src/dist/replication/meta_server/meta_split_service.cpp]: 所有写 zk 的优先级都应该是 high, 不需要在重试的时候改

**Comment by neverchanje** [comment]
include/dsn/cpp/serialization_helper/dsn.layer2_types.h
include/dsn/dist/replication/replication_types.h
src/dsn.layer2_types.cpp
src/dist/replication/client_lib/replication_types.cpp
这些文件都需要 format

File: src/dist/replication/meta_server/meta_split_service.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            }
+
+            auto &response = rpc.response();
+            response.err = ERR_OK;
+            response.partition_count = app->partition_count;
+
+        } else if (ec == ERR_TIMEOUT) {
+            dwarn_f("remote storage is not available now, please try it later");
+            tasking::enqueue(LPC_META_STATE_HIGH,
+                             nullptr,
```
Comment (neverchanje) [inline src/dist/replication/meta_server/meta_split_service.cpp]: remote storage 是用来省代码来让逻辑更清晰的. 你这里的 tracker 就是传 null, 是不安全的.

**Comment by hycdong** [comment]
刚刚交了一个新的commit，把code format都加上了 @neverchanje

**Comment by neverchanje** [comment]
哪个提交? 现在 diff 还是没 format

**Comment by hycdong** [comment]
commit 8a621626ddde23ec4f2724b60121624556c4cc4b

**Pull request closed without merging**